DIPs/DIP1xxx-JS.md

```diff
@@ -0,0 +1,86 @@
+# The Deprecation Process
+
+| Field | Value |
+|-----------------|-----------------------------------------------------------------|
+| DIP:            | xxxx |
+| Review Count:   | 0 |
+| Author:         | Jack Stouffer |
+| Implementation: | N/A |
+| Status:         | Draft |
+
+## Abstract
+
+In order to incrementally improve D or it's standard library, it's often necessary to
+mark features or functions for future removal. This document proposes a standardized
+process for language maintainers to remove public features. This process would be 
+used across DMD, Druntime, and Phobos.
+
+## Rationale
+
+There is general disagreement on the best and/or accepted way to remove public
+features. Each deprecation ends being handled slightly differently depending on
+who's handling the pull requests. Standardizing the process makes sure that
+deprecations are done very publicly and carefully, as to minimize breakage and
+to provide clear fixes for user code.
+
+## Description
+
+The following lays out the rules to follow to deprecate a part of D.
+
+A symbol or feature must not be marked for removal on a specific date, but rather on a
+specific release. This allows users to easily know if upgrading will break their
+code or not.
+
+At the time of the pull request for deprecation, all code in Phobos, Druntime,
+and DMD must be updated to remove use of the effected code. Any projects that
+are tested on the Project Tester that are broken should also have their
+maintainers notified.
+
+Both at the time of deprecation and removal, a changelog entry must be made. This
+changelog entry should have a short motivation for the deprecation (or removal)
+and should to describe what steps can be taken by the user to upgrade their codebase.
+
+In order to facilitate on schedule deprecations, a comment of the format
+`@@@DEPRECATED_[version]@@@` should be added to the top of the code to be removed/disabled.
+This comment allows code to be easily searched before every release to
+catch all planned deprecations.
+
+### Public Functions, Types, and Modules
+
+The symbol(s) must be marked using the `deprecated` keyword with a message containing
+the planned removal period and/or a pointer to more information pertaining to the
+deprecation. The documentation of the symbol(s) must be updated noting the
+deprecation and removal plan. The documentation should contain information to help
+the users using the symbol(s) transition their code away from the symbol(s).
+
+Users must be given at least four non-patch releases before the deprecated symbols
+are removed. More releases should be given if the removed code is commonly used.
+
+On the second release, the documentation for the symbol should be removed while
+keeping it public.
```

Comment:
This is way too fast. For years now, the policy has been to have a symbol deprecated but documented for a year and then deprecated and undocumented for a year. That has gotten complaints both that it is too quick and that it is too long, but it's tried to balance that. It's long enough that even projects that are rarely maintained have a decent chance to be updated but not so long that deprecated stuff is around forever.

As it stands, four major releases is only going to be something like 8 - 10 months - maybe 12 if we run into several delayed releases. And one release of being deprecated before being removed from the docs means that anyone who is not completely on top of things is just going to find the documentation missing the next time they rebuild their project. Even if we all agreed that the current process is too slow, I think that this is clearly too fast - especially when you consider that plenty of our users aren't constantly updating their compilers. Heck, with the documentation only being there for one major release, it'll probably be gone before the ldc guys update to that release, since they're frequently about a major release behind.

This should probably be more like 5 major releases as deprecated and documented and 5 as deprecated and undocumented with the caveat that it can be adjusted in either direction if deemed appropriate by the development team but that it should only be done so with very good reason (e.g. if having the symbol around is causing problems, it should probably go sooner, but that's rarely the case).

Reply:
That's a debate to have in the community review, methinks.